lib/jxl/dec_render_pipeline.h

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#ifndef LIB_JXL_DEC_RENDER_PIPELINE_H_
+#define LIB_JXL_DEC_RENDER_PIPELINE_H_
+
+#include <stdint.h>
+
+#include "lib/jxl/filters.h"
+
+namespace jxl {
+
+// The first pixel in the input to RenderPipelineStage will be located at
+// this position. Pixels before this position may be accessed as padding.
+constexpr size_t kRenderPipelineXOffset = 16;
+
+enum class RenderPipelineChannelMode {
+  kIgnored = 0,
+  kInPlace = 1,
+  kInOut = 2,
+};
+
+class RenderPipelineStage {
+ public:
+  // `input` points to `2*MaxPaddingY() + 1` pointers, each of which points to
+  // `3+num_non_color_channels` pointer-to-row. So, `input[MaxPaddingY()][0]` is
+  // the pointer to the center row of the first color channel.
+  //  `MaxPaddingY()` is the maximum value returned by `GetPaddingX()`;
+  //  typically, this is a constant.
+  // `output` points to `1<<MaxShiftY()` pointers, each of which points to
+  // `3+num_non_color_channels` pointer-to-row. So, `output[0][3]` is the
+  //  pointer to the top row of the first non-color channel.
+  //  `MaxShiftY()` is defined similarly to `MaxPaddingY()`.
+  //  `xsize` represents the total number of pixels to be processed in the input
+  //  row. `xpos` and `ypos` represent the position of the first pixel in the
+  //  center row in the input
+  virtual void ProcessRow(float* JXL_RESTRICT** input,
+                          float* JXL_RESTRICT** output, size_t xsize,
+                          size_t xpos, size_t ypos) const = 0;
+  virtual ~RenderPipelineStage() {}
+
+  // Amount of padding required by each channel in the various directions.
+  // The value for c=0 indicates padding required for color channels, subsequent
+  // values refer to padding for non-color channels, in order.
+  virtual size_t GetPaddingX(size_t c) const = 0;
+  virtual size_t GetPaddingY(size_t c) const = 0;
+
```

Comment:
I think this should be "border" not "padding". for me padding implies that the memory needs to be there but we don't use it (uninitialized is fine). I don't think we will ever need padding in Y.
Border in X and Y is often the same but it is ok to have two methods; padding in X would be a little bigger, but I think it could be BorderX rounded up to some vector size; anyway, what do you plan to use the padding (not border) information for? just that sum(all border + last padding) <= kRenderPipelineXOffset?

Reply:
You are correct, I meant border :)